packages/smithy-http/src/smithy_http/serializers.py

```diff
@@ -154,10 +183,22 @@ def begin_struct(self, schema: Schema) -> Iterator[ShapeSerializer]:
                     prefix=self._http_trait.query or "",
                 ),
             ),
-            fields=tuples_to_fields(headers),
+            fields=fields,
             body=payload,
         )
 
+    def _seek(self, payload: Any, pos: int, whence: int = 0) -> None:
+        if (
+            seek := getattr(payload, "seek", None)
+        ) is not None and not iscoroutinefunction(seek):
+            seek(pos, whence)
+
+    def _compute_content_length(self, payload: Any) -> int | None:
+        content_length = self._seek(payload, 0, 2)
+        if content_length is not None:
+            self._seek(payload, 0, 0)
+            return content_length
```

Comment:
I think this is a vector for data corruption right? If you get a partially read stream, we can't seek back to beginning. We'd want to use `tell` to check our current position before reading to the end or take the length and seek in reverse from the end. I think that's potentially going to have issues with length of bytes vs characters though.